### PR TITLE
Don't leak orphaned docs when deleting

### PR DESCRIPTION
**Pull request by gsoltis**
Previously, we weren't deleting our tracking of orphaned documents when we deleted the doc itself. I moved the deletion into the reference delegate, which is probably a better home anyways, and delete from the orphaned doc map at the same time so we no longer leak.

Caught while porting to web.

### CODE REVIEW

File: firebase-firestore/src/main/java/com/google/firebase/firestore/local/MemoryRemoteDocumentCache.java [hunk 1]
```diff
@@ -83,17 +83,7 @@ public MaybeDocument get(DocumentKey key) {
...
-    ImmutableSortedMap<DocumentKey, MaybeDocument> updated = docs;
-    for (Map.Entry<DocumentKey, MaybeDocument> entry : docs) {
-      if (!delegate.isPinned(entry.getKey(), upperBound)) {
-        updated = updated.remove(entry.getKey());
-        count++;
-      }
-    }
-    docs = updated;
-    return count;
+  ImmutableSortedMap<DocumentKey, MaybeDocument> getDocuments() {
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/local/MemoryRemoteDocumentCache.java]: Nit: Rename to `getAllDocuments()` to match `getAllDocumentsMatchingQuery()`.

Feel free to disagree :)

Comment (gsoltis) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/local/MemoryRemoteDocumentCache.java]: The other one should change to `getDocumentsMatchingQuery()`.

Comment (schmidt-sebastian) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/local/MemoryRemoteDocumentCache.java]: That's fine by me. Can you make this change (if you want, in a follow up?). `getDocumentsMatchingQuery()` is already the name we use on Web.

**Comment by gsoltis** [comment]
/test smoke-tests-experimental
--

**Pull request merged**